source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -377,6 +377,40 @@ absl::optional<SystemTime> SslSocket::expirationPeerCertificate() const {
   return Utility::getExpirationTime(*cert);
 }
 
+bool SslSocket::peerCertificateChainPresented() const {
+  bssl::UniquePtr<STACK_OF(X509)> cert_chain(SSL_get_peer_cert_chain(ssl_.get()));
+  return cert_chain != nullptr && sk_X509_num(cert_chain.get()) > 0;
+}
+
+const std::string& SslSocket::pemEncodedPeerCertificateChain() const {
+  if (!cached_pem_encoded_peer_cert_chain_.empty()) {
+    return cached_pem_encoded_peer_cert_chain_;
+  }
+
+  if (SSL_get_peer_cert_chain(ssl_.get()) == nullptr) {
+    ASSERT(cached_pem_encoded_peer_cert_chain_.empty());
+    return cached_pem_encoded_peer_cert_chain_;
+  }
+
+  bssl::UniquePtr<STACK_OF(X509)> cert_chain(
+      X509_chain_up_ref(SSL_get_peer_cert_chain(ssl_.get())));
```

Comment:
Do you need `X509_chain_up_ref` here? If you don't retain it then probably no need to.

Reply:
Addressed in d31dc1e.